### ISSUE TITLE
RNN / LSTM - typical training sensitivity to batchSize and initialisation environment / vars?

### ISSUE THREAD

**Issue by hsheil** [Element-Research/rnn#125]
I'm seeing two interesting behaviours when training the enhanced example code from #92 on the RecSys 2015 / YOOCHOOSE dataset (http://2015.recsyschallenge.com/challenge.html), viz:

1. Randomly, loss goes to 'nan' almost immediately after training begins.

2. Increasing batchSize > ~25 causes loss to move to either a huge or tiny -ve or +ve number (e.g. 1...e+34 or 1....e-34). This is not random and is guaranteed to happen after a certain quantile of batchSize.

The first issue is avoidable by simply restarting training - if the loss does not go to nan at the start of training then it remains well-behaved throughout training.

The second issue is avoidable only by keeping the batchSize < 25. As soon as this magic number is exceeded, the loss explodes / vanishes.

Are these two symptoms normal / expected and I should just have retry logic to detect the first and the second - just don't go for a large batchSize?

@nicholas-leonard I'll carry on training this code to see how it performs - a single epoch on my dev GPU (Titan Black) will take 9.9 hours (33 million events / 9.7 million sessions, batchSize = 20). I didn't want to push the code too early but if this training session delivers reasonable results (minimises the loss for the 5 features currently comprising the input), would be great to get your feedback on it before it goes into part two of the tutorial / real-world example.

**Comment by nicholas-leonard** [comment]
@hsheil Make sure you call zeroGradParameters() or gradParam:zero() before training begins as the gradParams could contain NaNs. Try smaller learning rates.

**Comment by hsheil** [comment]
@nicholas-leonard Will do, thanks for the tip! RE: learning rate, I've already reduced it down to 0.0001 but will experiment with it further as well.

**Comment by hsheil** [comment]
Hi @nicholas-leonard Just a quick update - I re-implemented the data loading code as discussed in #121 and now the network is resilient to pretty much all batch sizes (up to 896 - then I run out of memory). I suspect this behaviour was caused by a bug in my data loading code which I eliminated when I re-implemented getNext() - I'll rollback and check this at some point.

I selected the learning rate based on this paper (http://static.googleusercontent.com/media/research.google.com/en//pubs/archive/43905.pdf - pages 3/4) and it's also working well - all in all the overall characteristics are nicely well-behaved now after the latest set of changes.
